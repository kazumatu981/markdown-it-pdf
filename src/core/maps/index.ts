export * from './contents-map';
export * from './extension-map';
export * from './render-map';
export * from './contents-manager';
