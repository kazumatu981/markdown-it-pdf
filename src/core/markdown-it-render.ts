import MarkdownIt from 'markdown-it';
import fsPromises from 'fs/promises';

export interface Styles {
    internalUrls: string[];
    externalUrls: string[];
}

export class MarkdownItRender {
    private _md: MarkdownIt = new MarkdownIt();
    private _styles: Styles = {
        internalUrls: [],
        externalUrls: [],
    };

    /**
     * Adds the given URLs to the list of internal styles for this instance.
     *
     * @param {string[]} urls - The URLs of the styles to add.
     * @return {this} - Returns the current instance for method chaining.
     */
    public addStyles(urls: string[]): this {
        this._styles.internalUrls = urls;
        return this;
    }

    /**
     * Sets the external style URLs of this instance.
     *
     * @param {string[]} urls - The URLs to set as external styles.
     * @return {this} - Returns the current instance for method chaining.
     */
    public addExternalStyles(urls: string[]): this {
        this._styles.externalUrls = urls;
        return this;
    }

    /**
     * Uses the provided plugin with optional parameters.
     *
     * @param {MarkdownIt.PluginWithParams} plugin - The plugin to use.
     * @param {any[]} params - Optional parameters for the plugin.
     * @return {this} The MarkdownItRender instance for chaining.
     */
    public use(plugin: MarkdownIt.PluginWithParams, ...params: any[]): this {
        this._md.use(plugin, ...params);
        return this;
    }

    /**
     * Renders the given markdown string into HTML.
     *
     * @param {string} markdown - The markdown string to render.
     * @return {string} The rendered HTML string.
     */
    public render(markdown: string): string {
        const htmlBody = this._md.render(markdown);
        const styleTags = MarkdownItRender.generateStyleTags(
            this._styles.internalUrls,
            this._styles.externalUrls
        );
        return MarkdownItRender.htmlTemplate
            .replace(MarkdownItRender.styles, styleTags)
            .replace(MarkdownItRender.contents, htmlBody);
    }
    /**
     * Asynchronously renders a markdown file to HTML.
     *
     * @param {string} markdownFilePath - The path to the markdown file.
     * @return {Promise<string>} A promise that resolves to the rendered HTML.
     */
    public async renderFromFileAsync(
        markdownFilePath: string
    ): Promise<string> {
        const markdown = await fsPromises.readFile(markdownFilePath, 'utf8');
        return this.render(markdown);
    }

    //#region private members
    //#region Static Variables
    private static styles = '<!-- [[[STYLES]]] -->';
    private static contents = '<!-- [[[CONTENTS]]] -->';

    private static htmlTemplate = `
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    ${MarkdownItRender.styles}
</head>
<body>
    ${MarkdownItRender.contents}
</body>`;
    //#endregion

    //#region Static Methods
    /**
     * Generates style tag based on the given URL.
     * @param {string} stylePath css file path
     * @returns {string} generated style tag
     */
    private static generateStyleTag(stylePath: string): string {
        return `<link rel="stylesheet" type="text/css" href="${stylePath}">`;
    }
    /**
     * Generates style tags based on the given internal and external URLs.
     *
     * @param {string[]} internalUrls - An optional array of internal URLs.
     * @param {string[]} externalUrls - An optional array of external URLs.
     * @return {string} The generated style tags as a string.
     */
    private static generateStyleTags(
        internalUrls?: string[],
        externalUrls?: string[]
    ): string {
        return (
            (internalUrls?.map(MarkdownItRender.generateStyleTag).join('\n') ??
                '') +
            (externalUrls?.map(MarkdownItRender.generateStyleTag).join('\n') ??
                '')
        );
    }
    //#endregion
    //#endregion
}
