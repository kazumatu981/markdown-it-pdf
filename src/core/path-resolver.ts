import fs from 'fs';
import path from 'path';
import fsPromises from 'fs/promises';

type IsMatchFunction = (filePath: string) => boolean;
type TransformFunction<T> = (filePath: string) => T;

/**
 * Finds files in a given directory and returns an array of transformed results.
 *
 * @param {string} rootDir - The root directory to start the search from.
 * @param {boolean} recursive - Whether to search recursively in subdirectories.
 * @param {IsMatchFunction} isMatch - A function to check if a file matches a certain criteria.
 * @param {TransformFunction<T>} transform - A function to transform each found file.
 * @return {Promise<T[]>} A promise that resolves to an array of transformed file results.
 */
export async function findFiles<T>(
    rootDir: string,
    recursive: boolean,
    isMatch: IsMatchFunction,
    transform: TransformFunction<T>
): Promise<T[]> {
    const elements = await fsPromises.readdir(rootDir, { withFileTypes: true });
    const fileElements = elements
        .filter((e) => e.isFile() && isMatch(e.name))
        .map((e) => path.join(rootDir, e.name))
        .map(transform);
    if (recursive) {
        const subFileElements = await Promise.all(
            elements
                .filter((e) => e.isDirectory())
                .map((e) => path.join(rootDir, e.name))
                .map((d) => findFiles(d, recursive, isMatch, transform))
        );

        fileElements.push(...subFileElements.flat());
    }
    return fileElements;
}
/**
 * Converts a file path to a URL relative to the root directory.
 *
 * @param {string} rootDir - The root directory path.
 * @param {string} filePath - The file path to convert.
 * @return {string} The URL representing the file path relative to the root directory.
 * @throws {Error} If the file path is not in the root directory.
 */
export function filePathToUrl(rootDir: string, filePath: string): string {
    rootDir = path.resolve(process.cwd(), rootDir);
    filePath = path.resolve(process.cwd(), filePath);
    const relativePath = normalizePath(path.relative(rootDir, filePath));
    if (relativePath.startsWith('.')) {
        throw new Error('Path is not in rootDir');
    }
    return `/${relativePath}`;
}

/**
 * Normalizes a file path by converting it to a relative path and replacing any
 * Windows-style path separators with POSIX-style separators. Throws an error if
 * the file path is absolute.
 *
 * @param {string} filePath - The file path to normalize.
 * @return {string} The normalized file path.
 * @throws {Error} If the file path is absolute.
 */
export function normalizePath(filePath: string): string {
    if (path.isAbsolute(filePath)) {
        throw new Error('Path must be relative');
    }
    return path.normalize(filePath).split(path.win32.sep).join(path.posix.sep);
}

/**
 * Asynchronously builds a tree of directories based on the provided array of file paths.
 *
 * @param {string[]} files - An array of file paths.
 * @return {Promise<void>} A promise that resolves when the tree of directories is built successfully, or rejects with an error if there was a problem.
 */
export async function buildTreeOfFiles(files: string[]): Promise<void> {
    const dirs = Array.from(new Set(files.map((file) => path.dirname(file))));
    return new Promise((resolve, reject) => {
        Promise.all<void>(
            dirs.map(async (dir) => {
                await fsPromises.mkdir(dir, { recursive: true });
            })
        )
            .then((_) => resolve())
            .catch((err) => reject(err));
    });
}
