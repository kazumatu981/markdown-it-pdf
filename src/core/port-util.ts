import http from 'http';

interface Range {
    min?: number;
    max?: number;
}
interface SafeRange {
    min: number;
    max: number;
}
const privatePortRange: SafeRange = {
    min: 49152,
    max: 65535,
};

export interface ServerPort {
    port: number;
    httpServer: http.Server;
}

export interface ServerPortOptions {
    retry?: number;
    range?: Range;
}

function getRandom(range?: Range): number {
    const candidate = {
        min: range?.min ?? privatePortRange.min,
        max: range?.max ?? privatePortRange.max,
    };
    if (candidate.min > candidate.max) {
        throw new Error('Invalid range');
    }
    return Math.floor(
        Math.random() * (candidate.max - candidate.min) + candidate.min
    );
}

export async function tryToListen(
    port?: number,
    options?: ServerPortOptions
): Promise<ServerPort | undefined> {
    let serverPort = undefined;
    if (port != undefined) {
        serverPort = await tryToListenCore(port);
    } else {
        for (let i = 0; i < (options?.retry ?? 10); i++) {
            port = getRandom(options?.range);
            serverPort = await tryToListenCore(port);
            if (serverPort !== undefined) {
                break; // success
            }
        }
    }
    return serverPort;
}
function tryToListenCore(port: number): Promise<ServerPort | undefined> {
    const server = http.createServer();
    return new Promise((resolve) => {
        server.listen(port, () => {
            resolve({
                port,
                httpServer: server,
            });
        });
        server.on('error', () => {
            resolve(undefined);
        });
    });
}
