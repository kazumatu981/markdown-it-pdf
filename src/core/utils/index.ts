export * from './http-helper';
export * from './path-resolver';
