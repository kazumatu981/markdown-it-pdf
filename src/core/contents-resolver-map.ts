import { MarkdownItRender } from './markdown-it-render';
import fsPromises from 'fs/promises';
export type ContentType = 'markdown' | 'style';
export type ContentsResolverFunction = (filePath: string) => Promise<string>;

export class ContentsResolver extends Map<
    ContentType,
    ContentsResolverFunction
> {
    private _render: MarkdownItRender;
    public constructor(render: MarkdownItRender) {
        super();
        this._render = render;
        this.set('markdown', async (filePath) => {
            return this._render.renderFromFileAsync(filePath);
        });
        this.set('style', async (filePath) => {
            return fsPromises.readFile(filePath, 'utf8');
        });
    }

    public static resolveMediaType(contentType: ContentType): string{
        switch (contentType) {
            case 'markdown':
                return 'text/html';
            case 'style':
                return 'text/css';
            default:
                return 'text/plain';
        }
    }
    public async resolve(type: ContentType, filePath: string): Promise<string> {
        const resolver = this.get(type);
        if (!resolver) {
            throw new Error(`Unsupported contents type: ${type}`);
        }
        return resolver(filePath);
    }
}
