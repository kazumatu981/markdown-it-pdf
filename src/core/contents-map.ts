import { type MarkdownItRender } from './markdown-it-render';
import { findFiles, filePathToUrl } from './path-resolver';
import { type ResolverType, isSupported, getResolver } from './resolver-map';
import fsPromises from 'fs/promises';
import path from 'path';
export type ContentsResolverFunction = (
    filePath: string
) => Promise<string | Buffer>;

export interface ContentsResolver {
    mediaType: string;
    resolve: ContentsResolverFunction;
}

export interface ContentsMapEntity {
    url: string;
    resolverType: ResolverType;
    contentType: string;
    contentPath: string;
}

export interface RenderedEntity extends ContentsMapEntity {
    contents: string | Buffer;
}

export interface ContentsMapOptions {
    recursive?: boolean;
    externalUrls?: string[];
}
export class ContentsMap extends Map<string, ContentsMapEntity> {
    private _resolver: Map<ResolverType, ContentsResolverFunction>;
    private _render: MarkdownItRender;

    public static async createInstance(
        render: MarkdownItRender,
        contentsRoot: string,
        options?: ContentsMapOptions
    ): Promise<ContentsMap> {
        // create the instance
        const theInstance = new ContentsMap(render);

        // resolve to entity
        const contents = await findFiles<ContentsMapEntity>(
            contentsRoot,
            options?.recursive ?? true,
            isSupported,
            (filePath) => {
                const resolver = getResolver(path.extname(filePath));
                return {
                    url: filePathToUrl(contentsRoot, filePath),
                    resolverType: resolver.resolverType,
                    contentType:
                        resolver.resolvedContentType ?? resolver.contentType,
                    contentPath: filePath,
                };
            }
        );

        // set into entity
        contents.forEach((entry) => {
            theInstance.set(entry.url, entry);
        });

        // add into internalUrls
        const internalUrls = contents
            .filter((entry) => entry.contentType === 'text/css')
            .map((kv) => kv.url);
        render.addStyles(internalUrls);
        // add into externalUrls
        render.addExternalStyles(options?.externalUrls ?? []);

        return theInstance;
    }

    private constructor(render: MarkdownItRender) {
        super();
        this._render = render;
        this._resolver = this.generateDefaultResolvers();
    }

    public get markdownEntryUrls(): string[] {
        return [...this.keys()].filter(
            (url) => this.get(url)?.resolverType === 'markdown'
        );
    }

    public async render(filePath: string): Promise<RenderedEntity | undefined> {
        const entity = this.get(filePath);
        if (!entity) {
            return;
        }
        const contentResolver = this._resolver.get(entity.resolverType);
        if (!contentResolver) {
            return;
        }
        const contents = await contentResolver(entity.contentPath);
        return { ...entity, contents };
    }

    //#region private methods
    private generateDefaultResolvers(): Map<
        ResolverType,
        ContentsResolverFunction
    > {
        // add default resolvers
        return new Map<ResolverType, ContentsResolverFunction>()
            .set('markdown', this._markdownResolver.bind(this))
            .set('style', this._plainTextResolver.bind(this))
            .set('plainText', this._plainTextResolver.bind(this))
            .set('binary', this._binaryResolver.bind(this));
    }
    private async _markdownResolver(filePath: string): Promise<string> {
        return this._render.renderFromFileAsync(filePath);
    }
    private async _plainTextResolver(filePath: string): Promise<string> {
        return fsPromises.readFile(filePath, 'utf8');
    }
    private async _binaryResolver(filePath: string): Promise<Buffer> {
        return fsPromises.readFile(filePath);
    }
    //#endregion
}
