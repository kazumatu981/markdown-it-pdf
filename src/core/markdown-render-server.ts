import http from 'http';
import {
    type ContentsMapOptions,
    type RenderedEntity,
    RenderMap,
    ContentsMap,
} from './maps';
import {
    MarkdownItRender,
    utf8PlainTextRender,
    type HljsConfig,
} from './render';
import { type ListeningOptions, tryToListen } from './utils';
import { type Logger } from '../common';

/**
 * The options for the render server
 */
export interface RenderServerOptions
    extends ContentsMapOptions,
        ListeningOptions {
    /**
     * The port to listen on
     */
    port?: number;

    /**
     * The external urls for styles
     */
    externalUrls?: string[];
    /**
     * The template html file path
     */
    templatePath?: string;
    /**
     * The highlight.js config
     */
    hljs?: HljsConfig | false;
}

// TODO support user custom plugins on options.

/**
 * The markdown render server
 */
export class MarkdownRenderServer extends MarkdownItRender {
    private _options?: RenderServerOptions;
    private _server?: http.Server;
    private _contentsMap?: ContentsMap;
    private _listeningPort?: number;

    /**
     * Create a new markdown render server
     * @param rootDir {string} path to root directory of markdown files
     * @param options {RenderServerOptions} the options on this server launch.
     * @param logger {Logger} the logger
     * @returns {Promise<MarkdownRenderServer>} the markdown render server
     */
    public static async createInstance(
        rootDir?: string,
        options?: RenderServerOptions,
        logger?: Logger
    ): Promise<MarkdownRenderServer> {
        logger?.debug(
            `MarkdownRenderServer.createRenderServer() called with options: ${JSON.stringify(options)}`
        );

        // create this instance
        const theInstance = await new MarkdownRenderServer()
            // set the options
            .setOptions(options)
            // set the logger
            .setLogger(logger)
            // configure the template
            .configureTemplate({ ...options })
            // configure the contents map
            .then((instance) => {
                return instance.configureContentsMap(rootDir, options);
            })
            // refresh the contents map
            .then((instance) => {
                return instance.refresh(false);
            });
        logger?.debug(
            'contentsUrls: %o',
            theInstance.contentsMap.getEntityPaths()
        );

        return theInstance;
    }

    /**
     * Markdown paths which is available.
     * @returns {string[]} Markdown paths
     */
    public get availableMarkdownPaths(): string[] {
        return this.contentsMap.getEntityPaths('markdown');
    }

    /**
     * Style paths which is available.
     * @returns {string[]} Style paths
     */
    public get availableStylePaths(): string[] {
        return this.contentsMap.getEntityPaths('style');
    }

    /**
     * The url of the server. `http://localhost:{port}`
     * @returns {string} the url
     */
    public get myUrl(): string {
        return `http://localhost:${this._listeningPort}`;
    }

    private set contentsMap(contentsMap: ContentsMap) {
        this._contentsMap = contentsMap;
    }

    /**
     * The contents map
     * @returns {ContentsMap} the contents map
     */
    public get contentsMap(): ContentsMap {
        return this._contentsMap as ContentsMap;
    }

    /**
     * The listening port
     * @returns {number} the listening port
     */
    public get listeningPort(): number | undefined {
        return this._listeningPort;
    }

    /**
     * Start to listen from clients request.
     * @param port {number | undefined} The port to listen on
     * @returns {Promise<number>} A promise that resolves when the server is listening
     */
    public async listen(port?: number): Promise<number> {
        const portCandidate = port ?? this._options?.port;
        const serverPort = await tryToListen(
            portCandidate,
            this._options,
            this._logger
        );
        if (serverPort === undefined) {
            this._logger?.error('Can not listen on port %d', portCandidate);
            this._logger?.trace('Port not available');
            throw new Error('Can not listen');
        }
        this._listeningPort = serverPort.port;
        this._server = serverPort.httpServer;

        this._logger?.debug('Listening on port %d', this._listeningPort);

        // bind the server event listener
        this._server.on('request', this.onRequest.bind(this));
        return this._listeningPort;
    }

    /**
     * Close the server
     * @returns {Promise<void>} A promise that resolves when the server is closed
     */
    public async close(): Promise<void> {
        if (this._server === undefined) {
            return Promise.resolve();
        }
        this._listeningPort = undefined;
        return new Promise((resolve, rejects) => {
            (this._server as http.Server).close((err) => {
                if (err) {
                    this._logger?.error('Can not close server');
                    this._logger?.trace(err);
                    rejects(err);
                } else {
                    resolve();
                }
            });
        });
    }

    /**
     * Refresh the contents of the server
     * @param refreshContentsMap {boolean} whether to refresh the contents map
     * @returns {this} this instance, for method chaining
     */
    public async refresh(refreshContentsMap: boolean = true): Promise<this> {
        if (refreshContentsMap) {
            await this.contentsMap.refresh();
        }
        this.clearStyles();
        this.addStyles(this.availableStylePaths);
        this.clearExternalStyles();
        this.addExternalStyles(this._options?.externalUrls ?? []);
        return this;
    }

    //#region private methods

    //#region HTTP responses
    /**
     * Listener for incoming requests
     * @param req Incoming message
     * @param res Server response
     */
    private onRequest(
        req: http.IncomingMessage,
        res: http.ServerResponse
    ): void {
        this._logger?.debug('Request: %s', req.url);
        // Get requested file path
        const filePath = req.url as string;
        // Render the file contents with given path
        (this._contentsMap as ContentsMap).render(filePath).then((entity) => {
            // If file not found or rendering failed
            // Return "Not Found" to the client
            if (!entity) {
                this.writeNotFound(res);
                return;
            }
            // Write the rendered entity to the response
            this.writeEntity(res, entity);
        });
    }

    private writeNotFound(res: http.ServerResponse): void {
        this._logger?.warn(
            'Response for url: %s, status code: %d',
            res.req.url,
            404
        );
        res.statusCode = 404;
        res.appendHeader('Content-Type', 'text/plain');
        res.write('Not Found');
        res.end();
    }
    private writeEntity(
        res: http.ServerResponse,
        entity: RenderedEntity
    ): void {
        this._logger?.debug(
            'Response for url: %s, status code: %d',
            res.req.url,
            200
        );
        this._logger?.debug('content type: %s', entity.contentType);

        res.statusCode = 200;
        res.appendHeader('Content-Type', entity.contentType);
        res.write(entity.contents);
        res.end();
    }
    //#endregion

    //#region configuration functions
    private setOptions(options?: RenderServerOptions): this {
        this._options = options;
        return this;
    }
    private setLogger(logger?: Logger): this {
        this._logger = logger;
        return this;
    }
    private async configureContentsMap(
        rootDir?: string,
        options?: RenderServerOptions
    ): Promise<this> {
        const contentMap = await ContentsMap.createInstance(
            new RenderMap(),
            rootDir,
            options
        );
        contentMap.renderMap
            .set('markdown', this)
            .set('style', utf8PlainTextRender)
            .set('plainText', utf8PlainTextRender);
        this.contentsMap = contentMap;

        return this;
    }

    //#endregion

    //#endregion
}
