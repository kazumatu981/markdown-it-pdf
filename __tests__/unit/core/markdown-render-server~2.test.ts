import { jest, expect, describe, it } from '@jest/globals';
import { MarkdownRenderServer } from '../../../src/core/markdown-render-server';
import { mockingTestDir, unmockingTestDir } from '../../utils/test-dir';
import { readFromServer } from '../../utils/http-util';

import { mockLogger } from '../../utils/mock-logger';

describe('CoreLibrary Unit Tests - MarkdownRenderServer', () => {
    describe('serverListener', () => {
        afterEach(() => {
            jest.clearAllMocks();
        });
        it('can render md file', async () => {
            mockingTestDir();
            const server = await MarkdownRenderServer.createInstance('test');

            const port = await server.listen();

            const htmlData = await readFromServer(
                `http://localhost:${port}/test.md`
            );

            unmockingTestDir();
            expect(htmlData.statusCode).toEqual(200);
            expect(htmlData.contentType).toEqual('text/html');
            expect(htmlData).toMatchSnapshot('rendered md file');
            server.close();
        });
        it('can render md file - with logger', async () => {
            mockingTestDir();
            const server = await MarkdownRenderServer.createInstance(
                'test',
                {},
                mockLogger
            );

            const port = await server.listen(3001);

            const htmlData = await readFromServer(
                `http://localhost:${port}/test.md`
            );

            unmockingTestDir();
            expect(htmlData.statusCode).toEqual(200);
            expect(htmlData.contentType).toEqual('text/html');
            expect(htmlData).toMatchSnapshot('rendered md file');
            expect(mockLogger.debug).toMatchSnapshot();
            server.close();
        });
        it('can render style file', async () => {
            mockingTestDir();
            const server = await MarkdownRenderServer.createInstance('test');

            const port = await server.listen();

            const cssData = await readFromServer(
                `http://localhost:${port}/test.css`
            );

            unmockingTestDir();
            expect(cssData.statusCode).toEqual(200);
            expect(cssData.contentType).toEqual('text/css');
            expect(cssData).toMatchSnapshot('rendered css file');
            server.close();
        });
        it('can render txt file', async () => {
            mockingTestDir();
            const server = await MarkdownRenderServer.createInstance('test');

            const port = await server.listen();

            const txtData = await readFromServer(
                `http://localhost:${port}/sample.txt`
            );

            unmockingTestDir();
            expect(txtData.statusCode).toEqual(200);
            expect(txtData.contentType).toEqual('text/plain');
            expect(txtData).toMatchSnapshot('rendered txt file');
            server.close();
        });
        it('can render jpg file', async () => {
            mockingTestDir();
            const server = await MarkdownRenderServer.createInstance('test');

            const port = await server.listen();

            const txtData = await readFromServer(
                `http://localhost:${port}/sample.jpg`
            );

            unmockingTestDir();
            expect(txtData.statusCode).toEqual(200);
            expect(txtData.contentType).toEqual('image/jpeg');
            expect(txtData).toMatchSnapshot('rendered jpeg file');
            server.close();
        });
        it('root dir is not found', async () => {
            mockingTestDir();
            const server = await MarkdownRenderServer.createInstance('test');

            const port = await server.listen();

            const notFound = await readFromServer(`http://localhost:${port}`);

            unmockingTestDir();
            expect(notFound.statusCode).toEqual(404);
            expect(notFound.contentType).toEqual('text/plain');
            expect(notFound).toMatchSnapshot('Root is Not Found');
            server.close();
        });
        it('root dir is not found - with logger', async () => {
            mockingTestDir();
            const server = await MarkdownRenderServer.createInstance(
                'test',
                {},
                mockLogger
            );

            const port = await server.listen(3002);

            const notFound = await readFromServer(`http://localhost:${port}`);

            unmockingTestDir();
            expect(notFound.statusCode).toEqual(404);
            expect(notFound.contentType).toEqual('text/plain');
            expect(notFound).toMatchSnapshot('Root is Not Found');
            server.close();
            expect(mockLogger.warn).toMatchSnapshot();
        });
    });

    describe('fail to listen on retrying', () => {
        afterEach(() => {
            jest.clearAllMocks();
        });
        it('if port is already in use', async () => {
            mockingTestDir();
            const server0 = await MarkdownRenderServer.createInstance('test');
            const server1 = await MarkdownRenderServer.createInstance('test', {
                retry: 1,
                range: { min: 3500, max: 3500 },
            });

            const port0 = await server0.listen(3500);
            await expect(async () => {
                await server1.listen();
            }).rejects.toThrowError();
            await server0.close();
            unmockingTestDir();
        });
        it('if port is already in use - with logger', async () => {
            mockingTestDir();
            const server0 = await MarkdownRenderServer.createInstance('test');
            const server1 = await MarkdownRenderServer.createInstance(
                'test',
                {
                    retry: 1,
                    range: { min: 3500, max: 3500 },
                },
                mockLogger
            );

            const port0 = await server0.listen(3500);
            await expect(async () => {
                await server1.listen();
            }).rejects.toThrowError();
            await server0.close();
            unmockingTestDir();
            expect(mockLogger.error).toMatchSnapshot();
        });
    });
});
