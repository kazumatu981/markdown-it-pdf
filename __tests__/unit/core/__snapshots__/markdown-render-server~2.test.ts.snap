// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreLibrary Unit Tests - MarkdownRenderServer fail to listen on retrying if port is already in use - with logger 1`] = `
[MockFunction] {
  "calls": [
    [
      "Can not listen on port %d",
      undefined,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CoreLibrary Unit Tests - MarkdownRenderServer serverListener can render jpg file: rendered jpeg file 1`] = `
{
  "body": "sample.jpg",
  "contentType": "image/jpeg",
  "statusCode": 200,
}
`;

exports[`CoreLibrary Unit Tests - MarkdownRenderServer serverListener can render md file - with logger 1`] = `
[MockFunction] {
  "calls": [
    [
      "MarkdownRenderServer.createRenderServer() called with options: {}",
    ],
    [
      "template: %o",
      "
<!doctype html>
<html>
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <meta http-equiv="X-UA-Compatible" content="ie=edge" />
        <title>Document</title>
        {{#hljs}}
        <script src="{{js}}"></script>
        <link rel="stylesheet" type="text/css" href="{{css}}" />
        {{/hljs}}
        {{#styles}}
        <link rel="stylesheet" type="text/css" href="{{.}}" />
        {{/styles}}
    </head>
    <body>
        {{#hljs}}
        <script>hljs.highlightAll();</script>
        {{/hljs}}
        {{{body}}}
    </body>
</html>
",
    ],
    [
      "contentsUrls: %o",
      [
        "/sample.jpg",
        "/sample.txt",
        "/test.css",
        "/test.md",
        "/sub/sample.txt",
        "/sub/test.md",
        "/sub2/sample.txt",
        "/sub2/test.md",
        "/sub2/sub/sample.txt",
        "/sub2/sub/test.md",
      ],
    ],
    [
      "Try to Listen on port %d",
      3001,
    ],
    [
      "reserved port: %d",
      3001,
    ],
    [
      "Listening on port %d",
      3001,
    ],
    [
      "Request: %s",
      "/test.md",
    ],
    [
      "getModel() called.",
    ],
    [
      "styles: %o",
      [
        "/test.css",
      ],
    ],
    [
      "styles: %o",
      [],
    ],
    [
      "Response for url: %s, status code: %d",
      "/test.md",
      200,
    ],
    [
      "content type: %s",
      "text/html",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CoreLibrary Unit Tests - MarkdownRenderServer serverListener can render md file - with logger: rendered md file 1`] = `
{
  "body": "
<!doctype html>
<html>
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <meta http-equiv="X-UA-Compatible" content="ie=edge" />
        <title>Document</title>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.7.0/highlight.min.js"></script>
        <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.7.0/styles/github.min.css" />
        <link rel="stylesheet" type="text/css" href="/test.css" />
    </head>
    <body>
        <script>hljs.highlightAll();</script>
        <h1>test</h1>
<p>hello world</p>

    </body>
</html>
",
  "contentType": "text/html",
  "statusCode": 200,
}
`;

exports[`CoreLibrary Unit Tests - MarkdownRenderServer serverListener can render md file: rendered md file 1`] = `
{
  "body": "
<!doctype html>
<html>
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <meta http-equiv="X-UA-Compatible" content="ie=edge" />
        <title>Document</title>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.7.0/highlight.min.js"></script>
        <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.7.0/styles/github.min.css" />
        <link rel="stylesheet" type="text/css" href="/test.css" />
    </head>
    <body>
        <script>hljs.highlightAll();</script>
        <h1>test</h1>
<p>hello world</p>

    </body>
</html>
",
  "contentType": "text/html",
  "statusCode": 200,
}
`;

exports[`CoreLibrary Unit Tests - MarkdownRenderServer serverListener can render style file: rendered css file 1`] = `
{
  "body": "h1 { color: red; }",
  "contentType": "text/css",
  "statusCode": 200,
}
`;

exports[`CoreLibrary Unit Tests - MarkdownRenderServer serverListener can render txt file: rendered txt file 1`] = `
{
  "body": "hello world",
  "contentType": "text/plain",
  "statusCode": 200,
}
`;

exports[`CoreLibrary Unit Tests - MarkdownRenderServer serverListener root dir is not found - with logger 1`] = `
[MockFunction] {
  "calls": [
    [
      "Response for url: %s, status code: %d",
      "/",
      404,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CoreLibrary Unit Tests - MarkdownRenderServer serverListener root dir is not found - with logger: Root is Not Found 1`] = `
{
  "body": "Not Found",
  "contentType": "text/plain",
  "statusCode": 404,
}
`;

exports[`CoreLibrary Unit Tests - MarkdownRenderServer serverListener root dir is not found: Root is Not Found 1`] = `
{
  "body": "Not Found",
  "contentType": "text/plain",
  "statusCode": 404,
}
`;
